### DISCUSSION TITLE
Vercel Deployment for NextJS project (ReferenceError: document is not defined)

### DISCUSSION THREAD

**Discussion by codermaker2005** [vercel/next.js#69588] [Help]
### Summary

Hey, guys

So recently I've finished my NextJS project and headed to Vercel to deploy it. I'll provide the error message

Thanks in adavance

### Additional information

```js
[22:50:54.819] Running build in Washington, D.C., USA (East) – iad1
[22:50:55.037] Cloning github.com/codermaker2005/Student-Hub (Branch: main, Commit: 351799e)
[22:50:55.053] Skipping build cache, deployment was triggered without cache.
[22:50:55.693] Cloning completed: 653.035ms
[22:50:56.570] Running "vercel build"
[22:50:57.637] Vercel CLI 37.1.2
[22:50:58.357] Installing dependencies...
[22:51:01.246] npm warn deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
[22:51:01.875] npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
[22:51:02.902] npm warn deprecated @humanwhocodes/config-array@0.11.14: Use @eslint/config-array instead
[22:51:02.927] npm warn deprecated @humanwhocodes/object-schema@2.0.3: Use @eslint/object-schema instead
[22:51:02.976] npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
[22:51:09.207] 
[22:51:09.208] added 384 packages in 11s
[22:51:09.209] 
[22:51:09.209] 142 packages are looking for funding
[22:51:09.209]   run `npm fund` for details
[22:51:09.233] Detected Next.js version: 14.2.7
[22:51:09.237] Running "npm run build"
[22:51:09.391] 
[22:51:09.391] > studenthub2@0.1.0 build
[22:51:09.391] > next build
[22:51:09.391] 
[22:51:10.022] Attention: Next.js now collects completely anonymous telemetry regarding usage.
[22:51:10.023] This information is used to shape Next.js' roadmap and prioritize features.
[22:51:10.023] You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
[22:51:10.024] https://nextjs.org/telemetry
[22:51:10.024] 
[22:51:10.087]   ▲ Next.js 14.2.7
[22:51:10.088] 
[22:51:10.156]    Creating an optimized production build ...
[22:51:24.095]  ✓ Compiled successfully
[22:51:24.096]    Linting and checking validity of types ...
[22:51:25.825]    Collecting page data ...
[22:51:27.047]    Generating static pages (0/5) ...
[22:51:27.114]    Generating static pages (1/5) 
[22:51:27.308]    Generating static pages (2/5) 
[22:51:27.309]    Generating static pages (3/5) 
[22:51:27.403] ReferenceError: document is not defined
[22:51:27.404]     at o.current.o.current.window (/vercel/path0/.next/server/app/page.js:28:615)
[22:51:27.404]     at e1 (/vercel/path0/.next/server/app/page.js:28:1890)
[22:51:27.404]     at nj (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:46251)
[22:51:27.404]     at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47571)
[22:51:27.404]     at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64546)
[22:51:27.404]     at nI (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47010)
[22:51:27.404]     at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47717)
[22:51:27.404]     at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64546)
[22:51:27.404]     at nI (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47010)
[22:51:27.404]     at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47717) {
[22:51:27.404]   digest: '1236269061'
[22:51:27.404] }
[22:51:27.404] 
[22:51:27.404] Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
[22:51:27.404] 
[22:51:27.404] ReferenceError: document is not defined
[22:51:27.405]     at o.current.o.current.window (/vercel/path0/.next/server/app/page.js:28:615)
[22:51:27.405]     at e1 (/vercel/path0/.next/server/app/page.js:28:1890)
[22:51:27.405]     at nj (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:46251)
[22:51:27.405]     at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47571)
[22:51:27.405]     at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64546)
[22:51:27.405]     at nI (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47010)
[22:51:27.405]     at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47717)
[22:51:27.405]     at nN (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:64546)
[22:51:27.405]     at nI (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47010)
[22:51:27.405]     at nM (/vercel/path0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:47717)
[22:51:27.413]  ✓ Generating static pages (5/5)
[22:51:27.434] 
[22:51:27.434] > Export encountered errors on following paths:
[22:51:27.434] 	/page: /
[22:51:27.462] Error: Command "npm run build" exited with 1
[22:51:27.749]
```

This is my Repo in case it helped with something: https://github.com/codermaker2005/Student-Hub

**Comment by codermaker2005** [comment]
I asked ChatGPT perhaps it can provide a solution, but did not get that much


The error you're encountering during the Vercel build process is related to server-side rendering (SSR) in Next.js. Specifically, the error `ReferenceError: document is not defined` indicates that the `document` object is being used in a context where it is not available, which typically happens on the server-side. The `document` object is part of the DOM and is only available in the browser environment.

Here’s how you can address this issue:

### **1. Check Your Code for `document` Usage**

Ensure that any code accessing `document` is only run on the client-side. This can be done by wrapping such code in checks that confirm it's running in the browser:

```javascript
if (typeof window !== 'undefined') {
  // Code that uses `document` or `window` goes here
}
```

### **2. Use `useEffect` Hook for Client-Side Code**

If you’re using React hooks, make sure that code accessing `document` or `window` is inside `useEffect` or similar hooks that only run on the client-side:

```javascript
import { useEffect } from 'react';

const MyComponent = () => {
  useEffect(() => {
    // Client-side code here
    document.title = 'My Page Title';
  }, []);

  return <div>My Component</div>;
};
```

### **3. Check Next.js Configuration**

Ensure that your Next.js configuration is set up correctly for server-side rendering. Sometimes incorrect configurations or plugins can cause issues.

### **4. Ensure Dependencies Are Up-to-Date**

The warnings about deprecated packages suggest that some of your dependencies are outdated. Updating them might resolve some underlying issues:

```bash
npm update
```

### **5. Debug Locally**

Run the build process locally to see if you encounter the same issue. This can help you isolate whether the problem is with the Vercel environment or your code:

```bash
npm run build
```

### **6. Check for Issues in Your Codebase**

Examine the files being imported and used in the page causing the error. There might be components or functions trying to access `document` directly.

### **7. Review Documentation and Issues**

Check the [Next.js documentation](https://nextjs.org/docs) for details on server-side rendering and handling client-side code. You can also search for similar issues in the [Next.js GitHub repository](https://github.com/vercel/next.js/issues).

### **8. Deployment Configuration**

Ensure that your deployment configuration on Vercel matches the requirements of your Next.js application. This includes checking Node.js version settings and environment variables.

By addressing these points, you should be able to resolve the `document is not defined` error and successfully build and deploy your Next.js application on Vercel.

**Comment by Juneezee** [comment]
It seems to be caused by `react-router-dom`.  See https://github.com/colinhacks/nextjs-react-router and https://stackoverflow.com/a/75408439/7902371

Any particular reasons that you chose to use React Router instead of Next.js built-in router?

**Reply by codermaker2005** [reply]
Hey there, well actually I installed the package so I could route between pages in my navbar, I used it in my page.js. The reason is the routing at first was not working and when I was trying to navigate between pages I got 404, so installing the package had this issue gone.

**Reply by codermaker2005** [reply]
this is my repo in case you need it : https://github.com/codermaker2005/Student-Hub

**Comment by icyJoseph** [comment]
Hi,

In your repo you are making a dangerous bet, you are mixing routers. 

Routing using the History API is a bit of a make-belief thing. In short, it heavily relies on using the Link component a router provides, and it can cause really awkward behaviour when two routers are mixed.

In your case, the react-router-dom library, access the `document` API, at a top level, which means that just by sake of importing the library, you attempt to access the document object.

Now, `use client` doesn't mean that a component will defer to the client, rather, it means that the component will ship its JavaScript to the client, but it will SSR anyway. 

- https://github.com/reactwg/server-components/discussions/4

In your case, I'd say, the best thing you can do is use the Next.js File Based Router. 

You just need to create, for example, `app/internships/page.tsx`, and there:

```jsx
import Internships from '../internships';

export default function Page() {
  return <Internships />
}
```

And that's it. Then make sure to update Navbar to use the `next/link` component.

**Reply by codermaker2005** [reply]
Hey @icyJoseph thank you for your response. Yeah actually I'm still brand new to Next.js and trying to embark my journey. Recently I made progress and got rid of the error but got problems with navigation. You can check the discussion here: https://github.com/vercel/next.js/discussions/69617#discussioncomment-10537377

Thank you in advance for any guiding or help.

**Reply by icyJoseph** [reply]
Hi, ignore my previous answer, now deleted. I assumed you knew a bit more, but reading the entire thread I see you have a bunch of problems.

First, you need to move your navbar to `https://github.com/codermaker2005/Student-Hub/blob/main/app/layout.js`, that is, your layout component, and all of the stuff you have in `app/page.js`

Then in `app/page.js`, return `<Home />` as shown above for Internships.

Then you need to make sure you have `app/internships/page.js`, and do as shown above. Now you should be able to link from Home (/) to `/internships`

**Reply by codermaker2005** [reply]
thanks for the guidance @icyJoseph, so just to make sure and before applying the changes:

so the default file of `layout.js` is this: 

```
import { Inter } from "next/font/google";
import "./globals.css";

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body className={inter.className}>{children}</body>
    </html>
  );
}
```

I have to delete all the stuff in it and copy paste the code that's in the page.js to it, right?

second, in the `page.js`:

my code snippet is as follows:

```
"use client";

import React from 'react';
import { Analytics } from '@vercel/analytics/react';
import Head from 'next/head';  // Use next/head for managing the document head
import Navbar from './components/Navbar';
import GradesCalculator from './grades-calculator';  // Ensure correct import path
import Internships from './internships';  // Ensure correct import path
import Contact from './contact';  // Ensure correct import path
import Roadmaps from './roadmaps';  // Ensure correct import path
import Home from './home';  // Ensure correct import path


export default function App() {
  return (
    <>
      <Head>
        <title>Student Hub</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
     <Navbar/>
      <Home />

      <div className="relative min-h-screen bg-gray-900 text-white">
        {/* Dark background */}
        <div className="absolute inset-0 bg-gray-900"></div>
        {/* Gradient center */}
        <div className="absolute inset-0 flex justify-center items-center pointer-events-none">
          <div className="w-full h-full max-w-screen-lg max-h-screen "></div>
         
        </div>
      </div>
      <Analytics />
    </>
  );
}
```

it must be like that? 

```
import Home from './home';

export default function Page() {
  return <Home />
}
```

and lastly, as a file structure, if I got you right, I must create a folder for each component and inside each a file called `page.js` that got  this sort of code in my case? I mean like there's the `internships.jsx` normally in the `app/` as the other components, and I create a folder for each `page.js` in the` app/` too.
(as you said ) `app/internships/page.js`:

```
import Internships from '../internships';

export default function Page() {
  return <Internships />
}
```




` 

and this is my` internship.jsx` by the way:



```
import { useEffect, useState } from 'react';
import Papa from 'papaparse';
import Navbar from './components/Navbar';
export default function Internships() {
    const [internships, setInternships] = useState([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);

    useEffect(() => {
        const fetchInternships = async () => {
            try {
                const sheetUrl = 'https://docs.google.com/spreadsheets/d/1HVjbgjGm7F0YMWboLSkcGa2wj6TKSa4iFOZJPXI5jC4/pub?output=csv';
                const response = await fetch(sheetUrl);
                const csvData = await response.text();
                
                Papa.parse(csvData, {
                    header: true,
                    complete: (results) => {
                        setInternships(results.data);
                    },
                    error: (error) => {
                        setError(error.message);
                    }
                });
            } catch (error) {
                setError(error.message);
            } finally {
                setLoading(false);
            }
        };

        fetchInternships();
    }, []);

    return (
        <>
      <Navbar />
        <div className="min-h-screen bg-gray-900 text-white p-6">
            <h1 className="text-3xl font-semibold mb-8 text-center text-gray-100">
                Internship Opportunities
            </h1>
            {loading ? (
                <p className="text-center text-gray-400">Loading internships...</p>
            ) : error ? (
                <p className="text-center text-red-400">Error: {error}</p>
            ) : (
                <div className="max-w-5xl mx-auto grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6">
                    {internships.length === 0 ? (
                        <p className="text-center text-gray-400 col-span-full">No internships available at the moment.</p>
                    ) : (
                        internships.map((internship, index) => (
                            <div
                                key={index}
                                className="bg-gray-800 p-4 rounded-lg border border-gray-700 shadow-lg transform transition-transform duration-300 hover:scale-105 overflow-hidden flex flex-col justify-between"
                                style={{ minHeight: '200px' }}
                            >
                                <div className="flex-1">
                                    <h2 className="text-xl font-semibold mb-2 text-gray-100 truncate">{internship.Title}</h2>
                                    <p className="text-gray-300 overflow-hidden overflow-ellipsis">{internship.Description}</p>
                                </div>
                                {internship.Link && (
                                    <a href={internship.Link} className="block mt-4 text-blue-400 hover:text-blue-300">
                                        Learn More
                                    </a>
                                )}
                            </div>
                        ))
                    )}
                </div>
            )}
        </div>
        </>
    );
}

```


I apologize for the long response and for being a burden. And thank you again for your efforts!

**Reply by icyJoseph** [reply]
You put in, app/layout.js, everything that is shared by all pages.

And, yes, in App Router, pages are defined like, `app/about/page.js`

You could put the code directly on the page.js file too.

**Reply by icyJoseph** [reply]
Forgot to say that, when facing difficulties, it is probably better to take a step back to something smaller, and move forward with smaller steps, toward your goal.

**Reply by codermaker2005** [reply]
@icyJoseph i asked chatgpt to give me some insights and i guess it gave them right:

It seems you're trying to structure your Next.js project by organizing global settings and components shared across all pages. However, it looks like there are some issues and redundancies in your code. I'll help you clean it up and explain how it should be organized.

### 1. `app/layout.js` (Root Layout)

Your `RootLayout` should be responsible for wrapping all your pages with common layout elements like fonts, global styles, metadata, and possibly a persistent header or footer. Here’s a clean version:

```javascript
import { Inter } from 'next/font/google';
import './globals.css';

const inter = Inter({ subsets: ['latin'] });

export const metadata = {
  title: 'Student Hub',
  description: 'A platform to ease university students’ experiences with useful services',
};

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body className={inter.className}>
        {children}
      </body>
    </html>
  );
}
```

### 2. `pages/_app.js` (Application Component)

Next.js uses the `_app.js` file to initialize pages. This is where you should include global components that persist across page navigation, like your `Navbar`, and other components or providers. You don’t need to duplicate the metadata or layout elements here since they’re already handled by `RootLayout`.

Here’s how you can structure it:

```javascript
// pages/_app.js
import React from 'react';
import { Analytics } from '@vercel/analytics/react';
import Head from 'next/head';
import Navbar from '../components/Navbar';  // Adjusted import path

import '../styles/globals.css'; // Adjusted path for global styles

function MyApp({ Component, pageProps }) {
  return (
    <>
      <Head>
        <title>Student Hub</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <Navbar />
      <div className="relative min-h-screen bg-gray-900 text-white">
        <Component {...pageProps} />
      </div>
      <Analytics />
    </>
  );
}

export default MyApp;
```

### 3. `pages/index.js` (Home Page)

Your `Home` component should be placed here if it's the main content of your homepage:

```javascript
// pages/index.js
import Home from '../components/Home';  // Ensure this is the correct path

export default function IndexPage() {
  return (
    <div>
      <Home />
    </div>
  );
}
```

### 4. Organizing Components

Make sure your components like `Navbar`, `Home`, `GradesCalculator`, etc., are stored in a `components` directory and imported correctly in your pages. Ensure the import paths match where the components are located.

For example, if your `Navbar` component is in the `components` folder:

```javascript
// components/Navbar.js
import React from 'react';

export default function Navbar() {
  return (
    <nav>
      {/* Navbar content */}
    </nav>
  );
}
```

is that right?

**Reply by icyJoseph** [reply]
I'm gonna reserve my opinion about using such a tool. 

Please read the Next.js documentation, and try stuff on your own. 

- https://nextjs.org/docs/app/building-your-application/routing/defining-routes

It does look like you have sank a couple of days, using such tool, when a bit of reading and trying stuff on your own, would've gotten you way farther
